kube-runtime/src/scheduler.rs

```diff
@@ -126,6 +126,53 @@ impl<'a, T: Hash + Eq + Clone, R> SchedulerProj<'a, T, R> {
             }
         }
     }
+
+    /// Attempt to retrieve a message from queue and mark it as pending.
+    pub fn pop_queue_message_into_pending(&mut self, cx: &mut Context<'_>) -> Poll<T> {
+        loop {
+            match self.queue.poll_expired(cx) {
+                Poll::Ready(Some(msg)) => {
+                    let msg = msg.into_inner();
+                    self.pending.insert(msg);
+                }
+                Poll::Ready(None) | Poll::Pending => break Poll::Pending,
+            }
+        }
+    }
+}
+
+/// See [`Scheduler::hold`]
+pub struct Hold<'a, T, R> {
+    scheduler: Pin<&'a mut Scheduler<T, R>>,
+}
+
+impl<'a, T, R> Stream for Hold<'a, T, R>
+where
+    T: Eq + Hash + Clone,
+    R: Stream<Item = ScheduleRequest<T>>,
+{
+    type Item = T;
+
+    #[allow(clippy::match_wildcard_for_single_variants)]
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+        let mut scheduler = this.scheduler.as_mut().project();
+
+        loop {
+            match scheduler.requests.as_mut().poll_next(cx) {
+                Poll::Ready(Some(request)) => scheduler.schedule_message(request),
+                Poll::Ready(None) => return Poll::Ready(None),
+                Poll::Pending => break,
+            }
+        }
+
+        match scheduler.pop_queue_message_into_pending(cx) {
+            Poll::Pending => Poll::Pending,
+            // Since the above method never returns anything other than Poll::Pending
+            // we don't need to handle any other variant.
+            _ => unreachable!(),
+        }
```

Comment:
this does feel a bit awkward with how this one fn is "comment-depending" on behaviour there. if we don't intend to care about the return type of that fn, does it makes sense to call `scheduler.move_expired_queue_messages_into_pending()` and then simply return `Poll::Pending` without depending on the fn's return type?